lua/bufferline/events.lua

```diff
@@ -0,0 +1,248 @@
+
+      for tabnr, tabpage in ipairs(list_tabpages()) do
+        if not use_relative_file_paths or haslocaldir(-1, tabnr) == 1 then
+          use_relative_file_paths = false
+          break
+        end
+
+        for _, win in ipairs(tabpage_list_wins(tabpage)) do
+          if haslocaldir(win, tabnr) == 1 then
+            use_relative_file_paths = false
+            break
+          end
+        end
+      end
+
+      local buffers = {}
+      for _, bufnr in ipairs(state.buffers) do
+        local name = buf_get_name(bufnr)
+        if use_relative_file_paths then
+          name = utils.relative(name)
+        end
+
+        -- escape quotes
+        table_insert(buffers, {name = name, pinned =  state.get_buffer_data(bufnr).pinned})
+      end
+
+      vim.g.Bufferline__session_restore = "lua require'bufferline.state'.restore_buffers " ..
+        vim.inspect(buffers, {newline = ' ', indent = ''})
+    end,
+    group = augroup_misc,
+    pattern = 'SessionSavePre',
+  })
+
+  create_autocmd('WinClosed', {
+    callback = schedule_wrap(render.update),
+    group = augroup_render,
+  })
+
+  vim.schedule(function()
+    dictwatcherdel()
+    vim.cmd [[
+      " Must be global -_-
+      function! BufferlineOnOptionChanged(dict) abort
+        call luaeval("require'bufferline.events'.on_option_changed(_A)", a:dict)
+      endfunction
+
+      " Must be global -_-
+      function! BufferlineDictChanged(dict, key, changes) abort
+        silent! call dictwatcherdel(a:changes.old, '*', 'BufferlineOnOptionChanged')
+        call dictwatcheradd(a:changes.new, '*', 'BufferlineOnOptionChanged')
+        call BufferlineOnOptionChanged(a:changes.new)
+      endfunction
+
+      call dictwatcheradd(g:, 'bufferline', 'BufferlineDictChanged')
+      call dictwatcheradd(g:bufferline, '*', 'BufferlineOnOptionChanged')
+    ]]
+  end)
+
+  render.update()
+  enabled = true
+end
+
+--- What to do when the user configuration changes
+--- @param user_config table
+--- @return nil
+function events.on_option_changed(user_config)
+  options.setup(user_config) -- NOTE: must be first `setup` called here
+  highlight.setup()
+  JumpMode.set_letters(options.letters())
+
+  -- Don't jump-start barbar if it is disabled
+  if enabled then
+    render.update()
+  end
+end
+
+-- HACK: This is the only way to implement `dictwatcheradd` for global
+--       variables in Lua. The devs said in neovim/neovim#18393 and
+--       neovim/neovim#21469 that this is intended, no matter how hacky
+--       it is.
+do
+  --- The `vim.g` functionality
+  local vim_g_metatable = getmetatable(vim.g)
+
+  --- The `vim.g.foo = bar` function
+  local vim_g_metatable__newindex = vim_g_metatable.__newindex
+
+  --- @param tbl table
+  --- @param k string
+  --- @param v any
+  function vim_g_metatable.__newindex(tbl, k, v)
+    vim_g_metatable__newindex(tbl, k, v)
+    if k == 'bufferline' then
+      events.on_option_changed(v)
+    end
+  end
+end
```

Comment:
Neovim devs have been doing "interesting" choices in the last years.

Reply:
I think we could think about getting rid of the global options dict. I did it that way because I wrote this in vimscript first, and that's the way things were done back then. I find the new "setup" functions kinda lame, but they're a more or less established convention to setup plugin options now. We could eventually migrate to just exposing the setup function to configure barbar. I don't like breaking stuff, but we could do it gradually & clearly & add a notice ahead of time.

Reply:
> Neovim devs have been doing "interesting" choices in the last years.

A lot of stuff has been good, but I really disagree with encouraging users to rewrite parts of the `vim` stdlib in order to extend functonality, nor do I agree with the Lua-centrism of some decisions (e.g. `vim.notify` & `vim.paste`, using metatables instead of autocommands for dictwatchers, etc). 

I would have preferred setting a handler (e.g. `vim.lsp.handlers`) for certain editor activities rather than changing a the definition of functions every plugin depends on.

> We could eventually migrate to just exposing the `setup` function to configure barbar. I don't like breaking stuff, but we could do it gradually & clearly & add a notice ahead of time.

Yeah, the `g:bufferline` stuff is there for backwards-compatibility's sake, like you said. At this point, with the version we require, even users configuring from Vimscript can do `:lua require'bufferline'.setup{}`.

Removing `g:bufferline` would simplify a lot: no `dictwatcher`s, no hacky `vim.g` workarounds; and releasing a v2.0.0 would let us remove all the deprecated things that have been accumulated.

We can use the dictwatchers to notify people of `g:bufferline`'s deprecation. I'll open a v2.0.0 tracking issue to discuss any other changes.